keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2488 @@
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Definitions
+
+Policy definitions are responsible for:
+
+- Defining what validations the policy enforces and how violations are reported 
+- Defining how a policy may be configured
+
+Each `ValidatingAdmissionPolicy` resource may optionally set `spec.match`
+to constrain the resources it validates. 
+
+- `spec.match` constraints which resources this policy can be applied to. Policy
+  configurations each have match rules with further narrow this constraint, but
+  cannot expand it. This allows the CEL expressions to make safe assumptions.
+  E.g. a CEL expression that is constrained to CREATES and UPDATES of resources
+  is guaranteed the root `object` variable is never null, but a CEL expression
+  that might need to evaluate a DELETE must handle the root `object` variable
+  being null. See below "Match criteria" section for how match criteria is
+  described.
+- `spec.match` guides CEL expression type checking, see the below "Type safety"
+  section for more details.
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated 
+
+See below "Reporting violations to Clients" for more detail about how violations
+are reported.
+
+##### Policy Configurations
+
+Policy configurations are the interface between `ValidatingAdmissionPolicy`
+authors and the administrators that configure the policies for clusters. The
+configurations allows policy definition authors to define how exactly a
+`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
+schema.
+
+How these configuration CRDs are defined has not yet been decided.
+Alternatives are summarized here and discussed in more detail below:
+
+<<[UNRESOLVED jpbetz, liggitt, ?? ]>>
+Reach consensus on this.
+<<[/UNRESOLVED]>>
+
+| Design Alternative                                | Summary                        |
+| ------------------------------------------------- | ------------------------------ |
+| `PolicyConfiguration` kind with config reference  | Match criteria is defined in builtin type which references a config  | 
+| OpenAPIv3 `$ref` in CRDs                          | Match criteria is defined by k8s as a OpenAPIv3 type and CRDs schemas are modified to be allowed to reference it |
+| Duck Typed CRDs                                   | Require CRDs be defined with match field of correct structure |
```

Comment:
My initial impression is that I like this best.

Reply:
This remains my 1st choice as well. The main issue with this one is that the `spec.match` can be malformed (both in CRD schemas and in the data of CRs). We previewed this one with API review and got feedback that we should explore alternatives. I think I've now got a extensive list, and "`PolicyConfiguration` kind with config reference" is the only other viable contender from what I can tell (all other alternatives have _major_ problems).

I'll update this alternative with more detail about what the possible failure cases are and how they could be handled.